### PR TITLE
Advanced type hinting support

### LINKED ISSUE #78: Improve dataclass type hinting
### Is your feature request related to a problem? Please describe.

Our use of https://github.com/konradhalas/dacite is too primitive. It only supports to/from the object itself, not collections, optional, etc.

### Describe the solution you'd like

All of the following should be supported via type hints:

* `MyDataClass`
* `List[MyDataClass]`, `Set[MyDataClass]`, other collections, etc
* `Mapping[str, MyDataClass]`, `Dict[str, MyDataClass]`, other dicts, etc
* `Union[MyDataClass1, MyDataClass2]`, `Union[List[int], Dict[str, MyDataClass]]`, etc
* `Optional[MyDataClass1]`, `Optional[List[Dict[str, MyDataClass]]]`, etc

Here are the libraries found so far:

* https://github.com/konradhalas/dacite
  * What we currently use
  * No top-level collection, mapping, etc support
  * Could be good base to build off of
* https://github.com/lidatong/dataclasses-json
  * No support for non-decorator/non-mixin use (see issue on their repo numbered 243)
  * Requires marshmallow
* https://github.com/rnag/dataclass-wizard
  * Not the most popular (yet?)
  * Does not require inheritance/mixin use
  * Still requires us to figure out whether it's a dict or list or what not
* https://github.com/lovasoa/marshmallow_dataclass
  * Requires marshmallow (which is a popular library and therefore may clash with other users' marshmallow needs)
  * Seems to be the layer backing dataclasses-json
  * Will still have to determine data types expected

We should probably support customization via `instance.to_json()`/`instance.to_dict()` and `Class.to_json()`/`Class.to_dict()`. This means those using some of those other libraries will work automatically.

Needs investigation...

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

* Removed `dacite`
* Wrote custom type-hint support for deserializing

## Why?

Need to support all sorts of type hints like `Optional`, `Union`, etc

## Checklist

1. Closes #78

### CODE REVIEW

File: temporalio/converter.py [hunk 1]
```diff
@@ -776,3 +801,218 @@ def _type_hints_from_func(
             return (None, ret)
         args.append(arg_hint)
     return args, ret
+
+
+def value_to_type(hint: Type, value: Any) -> Any:
+    """Convert a given value to the given type hint.
+
+    This is used internally to convert a raw JSON loaded value to a specific
+    type hint.
```
Comment (Sushisource) [inline temporalio/converter.py:810]: Isn't this more like?
```suggestion
    This is used internally to convert a raw JSON loaded value to an instance
    of a specific type.
```

**Review by Sushisource — APPROVED** [review]
Looks like that wasn't too bad

File: temporalio/converter.py (hunk 1 above)
Comment (cretz) [reply temporalio/converter.py:810]:
```suggestion
    This is used internally to convert a raw JSON loaded value using a specific
    type hint.
```

That may be clearer. Will change if there is another commit here.

**Comment by cretz** [comment]
> Looks like that wasn't too bad

@Sushisource - It wasn't too bad, but I fear people may have custom needs and it may be hard to document in detailed form what we specifically support. Also, while the code is small, it may not be obvious to users that, for example, a union tries all union types and chooses the first one (so if the first dataclass matched I wouldn't try the second).

**Review by bergundy — APPROVED** [review]
Tests look good didn't look into the implementation too much.

**Pull request merged**